File: sdks/python/hatchet_sdk/features/cel.py
import asyncio
from typing import Literal

from pydantic import BaseModel, Field

from hatchet_sdk.clients.rest.api.cel_api import CELApi
from hatchet_sdk.clients.rest.api_client import ApiClient
from hatchet_sdk.clients.rest.models.v1_cel_debug_request import V1CELDebugRequest
from hatchet_sdk.clients.rest.models.v1_cel_debug_response_status import (
    V1CELDebugResponseStatus,
)
from hatchet_sdk.clients.v1.api_client import BaseRestClient, retry
from hatchet_sdk.utils.typing import JSONSerializableMapping


class CELSuccess(BaseModel):
    status: Literal["success"] = "success"
    output: bool


class CELFailure(BaseModel):
    status: Literal["failure"] = "failure"
    error: str


class CELEvaluationResult(BaseModel):
    result: CELSuccess | CELFailure = Field(discriminator="status")


class CELClient(BaseRestClient):
    """
    The CEL client is a client for debugging CEL expressions within Hatchet
    """

    def _ca(self, client: ApiClient) -> CELApi:
        return CELApi(client)

    @retry
    def debug(
        self,
        expression: str,
        input: JSONSerializableMapping,
        additional_metadata: JSONSerializableMapping | None = None,
        filter_payload: JSONSerializableMapping | None = None,
    ) -> CELEvaluationResult:
        """
        Debug a CEL expression with the provided input, filter payload, and optional metadata. Useful for testing and validating CEL expressions and debugging issues in production.

        :param expression: The CEL expression to debug.
        :param input: The input, which simulates the workflow run input.
        :param additional_metadata: Additional metadata, which simulates metadata that could be sent with an event or a workflow run
        :param filter_payload: The filter payload, which simulates a payload set on a previous-created filter

        :raises ValueError: If no response is received from the CEL debug API.

        :return: A V1CELDebugErrorResponse or V1CELDebugSuccessResponse containing the result of the debug operation.
        """
        request = V1CELDebugRequest(
            expression=expression,
            input=input,
            additionalMetadata=additional_metadata,
            filterPayload=filter_payload,
        )
        with self.client() as client:
            result = self._ca(client).v1_cel_debug(
                tenant=self.client_config.tenant_id, v1_cel_debug_request=request
            )

            if result.status == V1CELDebugResponseStatus.ERROR:
                if result.error is None:
                    raise ValueError("No error message received from CEL debug API.")

                return CELEvaluationResult(result=CELFailure(error=result.error))

            if result.output is None:
                raise ValueError("No output received from CEL debug API.")

            return CELEvaluationResult(result=CELSuccess(output=result.output))

    async def aio_debug(
        self,
        expression: str,
        input: JSONSerializableMapping,
        additional_metadata: JSONSerializableMapping | None = None,
        filter_payload: JSONSerializableMapping | None = None,
    ) -> CELEvaluationResult:
        """
        Debug a CEL expression with the provided input, filter payload, and optional metadata. Useful for testing and validating CEL expressions and debugging issues in production.

        :param expression: The CEL expression to debug.
        :param input: The input, which simulates the workflow run input.
        :param additional_metadata: Additional metadata, which simulates metadata that could be sent with an event or a workflow run
        :param filter_payload: The filter payload, which simulates a payload set on a previous-created filter

        :return: A V1CELDebugErrorResponse or V1CELDebugSuccessResponse containing the result of the debug operation.
        """
        return await asyncio.to_thread(
            self.debug, expression, input, additional_metadata, filter_payload
        )

File: sdks/ruby/src/lib/hatchet/features/cel.rb
# frozen_string_literal: true

module Hatchet
  module Features
    # Result types for CEL expression evaluation
    CELSuccess = Struct.new(:status, :output, keyword_init: true) do
      def initialize(output:)
        super(status: 'success', output: output)
      end
    end

    CELFailure = Struct.new(:status, :error, keyword_init: true) do
      def initialize(error:)
        super(status: 'failure', error: error)
      end
    end

    CELEvaluationResult = Struct.new(:result, keyword_init: true)

    # CEL client for debugging CEL expressions within Hatchet
    #
    # This class provides a high-level interface for testing and debugging
    # CEL (Common Expression Language) expressions used in filters and conditions.
    #
    # @example Debugging a CEL expression
    #   result = cel_client.debug(
    #     expression: 'input.value > 10',
    #     input: { value: 15 }
    #   )
    #   if result.result.status == 'success'
    #     puts "Output: #{result.result.output}"
    #   else
    #     puts "Error: #{result.result.error}"
    #   end
    #
    # @since 0.1.0
    class CEL
      # Initializes a new CEL client instance
      #
      # @param rest_client [Object] The configured REST client for API communication
      # @param config [Hatchet::Config] The Hatchet configuration containing tenant_id and other settings
      # @return [void]
      # @since 0.1.0
      def initialize(rest_client, config)
        @rest_client = rest_client
        @config = config
        @cel_api = HatchetSdkRest::CELApi.new(rest_client)
      end

      # Debug a CEL expression with provided input and optional metadata
      #
      # Useful for testing and validating CEL expressions and debugging issues in production.
      #
      # @param expression [String] The CEL expression to debug
      # @param input [Hash] The input, which simulates the workflow run input
      # @param additional_metadata [Hash, nil] Additional metadata simulating metadata sent with an event or workflow run
      # @param filter_payload [Hash, nil] The filter payload simulating a payload set on a previously-created filter
      # @return [CELEvaluationResult] A result containing either a CELSuccess or CELFailure
      # @raise [RuntimeError] If no response is received from the CEL debug API
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   result = cel_client.debug(
      #     expression: 'input.count > 5 && metadata.env == "prod"',
      #     input: { count: 10 },
      #     additional_metadata: { env: 'prod' }
      #   )
      def debug(expression:, input:, additional_metadata: nil, filter_payload: nil)
        request = HatchetSdkRest::V1CELDebugRequest.new(
          expression: expression,
          input: input,
          additional_metadata: additional_metadata,
          filter_payload: filter_payload
        )

        result = @cel_api.v1_cel_debug(@config.tenant_id, request)

        if result.status == 'ERROR' || result.status == :ERROR
          raise 'No error message received from CEL debug API.' if result.error.nil?

          return CELEvaluationResult.new(result: CELFailure.new(error: result.error))
        end

        raise 'No output received from CEL debug API.' if result.output.nil?

        CELEvaluationResult.new(result: CELSuccess.new(output: result.output))
      end
    end
  end
end
